src/scriptrag/api/database_operations.py

```diff
@@ -0,0 +1,499 @@
+            scene_id: ID of the parent scene
+            dialogues: List of dialogue objects
+            character_map: Mapping of character names to IDs
+
+        Returns:
+            Number of dialogues inserted
+        """
+        count = 0
+        for order, dialogue in enumerate(dialogues):
+            if dialogue.character not in character_map:
+                # Skip dialogues for unknown characters
+                logger.warning(f"Unknown character in dialogue: {dialogue.character}")
+                continue
+
+            metadata = {}
+            if dialogue.parenthetical:
+                metadata["parenthetical"] = dialogue.parenthetical
+
+            conn.execute(
+                """
+                INSERT INTO dialogues (scene_id, character_id, dialogue_text,
+                                     order_in_scene, metadata)
+                VALUES (?, ?, ?, ?, ?)
+                """,
+                (
+                    scene_id,
+                    character_map[dialogue.character],
+                    dialogue.text,
+                    order,
+                    json.dumps(metadata) if metadata else None,
+                ),
+            )
+            count += 1
+
+        logger.debug(f"Inserted {count} dialogues for scene {scene_id}")
+        return count
+
+    def insert_actions(
+        self, conn: sqlite3.Connection, scene_id: int, actions: list[str]
+    ) -> int:
+        """Insert action records.
+
+        Args:
+            conn: Database connection
+            scene_id: ID of the parent scene
+            actions: List of action text lines
+
+        Returns:
+            Number of actions inserted
+        """
+        count = 0
+        for order, action_text in enumerate(actions):
+            if not action_text.strip():
+                continue
+
+            conn.execute(
+                """
+                INSERT INTO actions (scene_id, action_text, order_in_scene)
+                VALUES (?, ?, ?)
+                """,
+                (scene_id, action_text, order),
+            )
+            count += 1
+
+        logger.debug(f"Inserted {count} actions for scene {scene_id}")
+        return count
+
+    def _extract_location(self, heading: str) -> str | None:
+        """Extract location from scene heading.
+
+        Args:
+            heading: Scene heading text
+
+        Returns:
+            Extracted location or None
+        """
+        # Basic extraction - can be enhanced
+        parts = heading.split(" - ")
+        if len(parts) > 0:
+            # Remove INT./EXT. prefix
+            location = parts[0]
+            for prefix in ["INT.", "EXT.", "INT", "EXT", "I/E"]:
+                if location.startswith(prefix):
+                    location = location[len(prefix) :].strip()
+                    break
+            return location if location else None
+        return None
+
+    def _extract_time(self, heading: str) -> str | None:
```

Comment:
Same here. Check if this is code duplication, and even if not move it to ScreenplayUtils